src/libchdr_chd.c

```diff
@@ -712,22 +714,38 @@ static chd_error cdlz_codec_decompress(void *codec, const uint8_t *src, uint32_t
 {
 	uint32_t framenum;
 	cdlz_codec_data* cdlz = (cdlz_codec_data*)codec;
+	chd_error decomp_err;
 
 	/* determine header bytes */
-	uint32_t frames = destlen / CD_FRAME_SIZE;
-	uint32_t complen_bytes = (destlen < 65536) ? 2 : 3;
-	uint32_t ecc_bytes = (frames + 7) / 8;
-	uint32_t header_bytes = ecc_bytes + complen_bytes;
+	const uint32_t frames = destlen / CD_FRAME_SIZE;
+	const uint32_t complen_bytes = (destlen < 65536) ? 2 : 3;
+	const uint32_t ecc_bytes = (frames + 7) / 8;
+	const uint32_t header_bytes = ecc_bytes + complen_bytes;
+
+	/* input may be truncated, double-check */
+	if ((ecc_bytes + 2) > complen)
+		return CHDERR_DECOMPRESSION_ERROR;
 
 	/* extract compressed length of base */
 	uint32_t complen_base = (src[ecc_bytes + 0] << 8) | src[ecc_bytes + 1];
```

Comment:
There's an inconsistency in variable declarations here; Your new code declares variables at the start of functions, but here (and in other places) a variable is declared in between statements.
If the latter is acceptable, then perhaps declare the error variables only when they are first assigned to?

Reply:
I'm all for declare-at-time-of-use. If you don't care about C89, then I'd absolutely move it down.
Actually, having the bounds check in there would probably break C89 anyway, so I'd have to move `complen_base` up.. ugh.